photon-ml/src/main/scala/com/linkedin/photon/ml/algorithm/CoordinateDescent.scala

```diff
@@ -177,36 +188,48 @@ class CoordinateDescent(
-        validatingScoresContainerOption = validatingDataAndEvaluatorsOption.map { case (validatingData, evaluators) =>
+        // Update the validation score and evaluate the updated model on the validation data
+        // if there are evaluators defined
+        val currentEval = validatingDataAndEvaluatorsOption.map { case (validatingData, evaluators) =>
           var validatingScoresContainer = validatingScoresContainerOption.get
-          val validatingScores = updatedModel
-            .score(validatingData)
-            .setName(s"Updated validating scores with coordinateId $coordinateId")
-            .persistRDD(StorageLevel.INFREQUENT_REUSE_RDD_STORAGE_LEVEL)
-            .materialize()
+          validatingScoresContainer = validatingScoresContainer.updated(coordinateId,
+            updatedModel
+              .score(validatingData)
+              .setName(s"Updated validating scores for coordinateId $coordinateId")
+              .persistRDD(StorageLevel.INFREQUENT_REUSE_RDD_STORAGE_LEVEL)
+              .materialize()
+          )
           validatingScoresContainer(coordinateId).unpersistRDD()
-          validatingScoresContainer = validatingScoresContainer.updated(coordinateId, validatingScores)
           val fullScore = validatingScoresContainer.values.reduce(_ + _)
-          evaluators.foreach { evaluator =>
+          evaluators.map { evaluator =>
             val validationTimer = Timer.start()
             val evaluationMetric = evaluator.evaluate(fullScore.scores)
-            logger.debug(s"Finished validating the model, time elapsed: ${validationTimer.stop().durationSeconds} (s).")
+            logger.debug(s"Finished model validation, time elapsed: ${validationTimer.stop().durationSeconds} (s).")
             logger.info(s"Evaluation metric computed with ${evaluator.getEvaluatorName} after updating " +
-                s"coordinateId $coordinateId at iteration $iteration is $evaluationMetric")
-          }
-          validatingScoresContainer
+              s"coordinateId $coordinateId at iteration $iteration is: $evaluationMetric")
+            evaluationMetric
+          }.head // the first evaluator is the one we use for hyper-parameter optimization
         }
 
-        logger.info(
-          s"Updating coordinate $coordinateId finished, time elapsed: ${coordinateTimer.stop().durationSeconds} (s)\n")
-      }
+        updatedGAMEModel = updatedGAMEModel.updateModel(coordinateId, updatedModel)
 
-      logger.info(
-        s"Iteration $iteration of coordinate descent finished, time elapsed: " +
-        s"${iterationTimer.stop().durationSeconds} (s)\n\n")
-    }
+        // The "best" model is whichever one optimizes the value of the first evaluation function
+        if (bestEval.isEmpty || currentEval.get > bestEval.get) {
+          bestEval = currentEval
+          bestModel = Some(updatedGAMEModel)
+          logger.debug(s"Found better GAME model, with evaluation: $currentEval")
+        }
```

Comment:
Hmm - that's a good point. I definitely lost sight of that. It shouldn't be selecting coordinate models in the middle of an iteration, but do you object to recording mid-iteration GAME models as the best one as well? I'm asking because you bring up waiting until all coordinates in an outer-iteration have been updated.